lib/lf/geometry/segment_o2.cc

```diff
@@ -0,0 +1,171 @@
+/**
+ * @file
+ * @brief Implementation of second-order segments
+ * @author Anian Ruoss
+ * @date   2018-11-18 19:02:17
+ * @copyright MIT License
+ */
+
+#include "segment_o2.h"
+#include "point.h"
+
+namespace lf::geometry {
+
+Eigen::MatrixXd SegmentO2::Global(const Eigen::MatrixXd& local) const {
+  Eigen::MatrixXd global(DimGlobal(), local.cols());
+
+  const Eigen::VectorXd vtx0 = coords_.col(0);
+  const Eigen::VectorXd vtx1 = coords_.col(1);
+  const Eigen::VectorXd midp = coords_.col(2);
+
+  // polynomial of degree 2: alpha * x^2 + beta * x + gamma
+  const Eigen::VectorXd alpha = 2. * (vtx1 + vtx0) - 4. * midp;
+  const Eigen::VectorXd beta = 4. * midp - 3. * vtx0 - vtx1;
+  const Eigen::VectorXd& gamma = vtx0;
+
+  for (size_t point = 0; point < local.cols(); ++point) {
+    const double x = local(point);
+
+    if (0. <= x && x <= 1.) {
+      global.col(point) = alpha * x * x + beta * x + gamma;
+    } else {
+      LF_VERIFY_MSG(false,
+                    "local coordinate out of bounds for reference element");
+    }
+  }
+
+  return global;
+}
+
+Eigen::MatrixXd SegmentO2::Jacobian(const Eigen::MatrixXd& local) const {
+  Eigen::MatrixXd jacobian(DimGlobal(), DimLocal() * local.cols());
+
+  const Eigen::VectorXd vtx0 = coords_.col(0);
+  const Eigen::VectorXd vtx1 = coords_.col(1);
+  const Eigen::VectorXd midp = coords_.col(2);
+
+  // polynomial of degree 2: alpha * x^2 + beta * x + gamma
+  const Eigen::VectorXd alpha = 2. * (vtx1 + vtx0) - 4. * midp;
+  const Eigen::VectorXd beta = 4. * midp - 3. * vtx0 - vtx1;
+
+  for (size_t point = 0; point < local.cols(); ++point) {
+    const double x = local(point);
+
+    if (0. <= x && x <= 1.) {
+      jacobian.col(point) = 2. * alpha * x + beta;
+    } else {
+      LF_VERIFY_MSG(false,
+                    "local coordinate out of bounds for reference element");
+    }
+  }
+
+  return jacobian;
+}
+
+Eigen::MatrixXd SegmentO2::JacobianInverseGramian(
+    const ::Eigen::MatrixXd& local) const {
+  Eigen::MatrixXd jacInvGram(DimGlobal(), DimLocal() * local.cols());
+  Eigen::MatrixXd jacobian = Jacobian(local);
+
+  for (size_t point = 0; point < local.cols(); ++point) {
+    Eigen::MatrixXd jac =
+        jacobian.block(0, point * DimLocal(), DimGlobal(), DimLocal());
+
+    if (DimGlobal() == DimLocal()) {
+      jacInvGram.block(0, point * DimLocal(), DimGlobal(), DimLocal()) =
+          jac.inverse().transpose();
+    } else {
+      jacInvGram.block(0, point * DimLocal(), DimGlobal(), DimLocal()) =
+          jac * (jac.transpose() * jac).inverse();
+    }
+  }
+
+  return jacInvGram;
+}
+
+Eigen::VectorXd SegmentO2::IntegrationElement(
```

Comment:
The Gramian determinant is a polynomial of degree 4, whose coefficient can be precomputed in the constructor and reused for all invocations of IntegrationElement() and JacobianInverseGramian().